File: code-tiny/tools/web_framework/pipeline.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Sequence, Tuple

from .models import EndpointFact, WebAnalysisResult, stable_id


_SKIP_DIRS = frozenset({
    ".git", ".cache", ".venv", "venv", "node_modules", "vendor", "dist", "build",
    "__pycache__", "bin", "obj",
})
_FASTAPI_RE = re.compile(
    r"@(?:\w+\.)?(?:app|router|api_router|blueprint)\.(?P<method>get|post|put|delete|patch|head|options|route|websocket)"
    r"\s*\(\s*[rRuUfF]*['\"](?P<path>[^'\"]+)['\"][^)]*\)\s*"
    r"(?:async\s+)?def\s+(?P<handler>[A-Za-z_]\w*)\s*\(",
    re.IGNORECASE | re.DOTALL,
)
_DJANGO_RE = re.compile(
    r"\b(?:path|re_path)\s*\(\s*[rRuU]*['\"](?P<path>[^'\"]+)['\"]\s*,\s*"
    r"(?P<handler>[A-Za-z_]\w*(?:\.[A-Za-z_]\w*)*)(?:\.as_view\s*\(\s*\))?",
    re.MULTILINE,
)
_EXPRESS_RE = re.compile(
    r"\b(?:app|router|server|api)\.(?P<method>get|post|put|delete|patch|head|options|all|use)"
    r"\s*\(\s*['\"](?P<path>[^'\"]+)['\"]\s*,\s*(?P<handler>[A-Za-z_$][\w$]*)",
    re.IGNORECASE,
)
_LARAVEL_RE = re.compile(
    r"\bRoute::(?P<method>get|post|put|delete|patch|options|any|match)\s*\(\s*"
    r"['\"](?P<path>[^'\"]+)['\"]\s*,\s*\[\s*(?P<scope>[A-Za-z_\\][\w\\]*)::class"
    r"\s*,\s*['\"](?P<handler>[A-Za-z_]\w*)['\"]\s*\]",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class _Symbol:
    name: str
    scope: str
    file_path: str
    label: str = "Function"


def _relative(root: Path, path: Path) -> str:
    return path.relative_to(root).as_posix()


def _source_files(root: Path) -> Iterable[Path]:
    for path in root.rglob("*"):
        if not path.is_file() or any(part in _SKIP_DIRS for part in path.relative_to(root).parts):
            continue
        if path.suffix.lower() in {".py", ".js", ".jsx", ".php"}:
            yield path


def _line(text: str, offset: int) -> int:
    return text.count("\n", 0, offset) + 1


def _normalize_path(value: str) -> str:
    path = value.strip()
    if not path.startswith("/"):
        path = "/" + path
    return re.sub(r"/{2,}", "/", path)


def _symbol_index(root: Path, files: Sequence[Path]) -> Dict[str, List[_Symbol]]:
    index: Dict[str, List[_Symbol]] = {}
    for path in files:
        text = path.read_text(encoding="utf-8", errors="ignore")
        rel = _relative(root, path)
        suffix = path.suffix.lower()
        if suffix == ".py":
            for match in re.finditer(r"(?m)^\s*(?:async\s+)?def\s+([A-Za-z_]\w*)\s*\(", text):
                symbol = _Symbol(match.group(1), "", rel)
                index.setdefault(symbol.name, []).append(symbol)
        elif suffix in {".js", ".jsx"}:
            patterns = (
                r"(?m)^\s*(?:export\s+)?(?:async\s+)?function\s+([A-Za-z_$][\w$]*)\s*\(",
                r"(?m)^\s*(?:export\s+)?(?:const|let|var)\s+([A-Za-z_$][\w$]*)\s*=\s*(?:async\s*)?\([^)]*\)\s*=>",
            )
            for pattern in patterns:
                for match in re.finditer(pattern, text):
                    symbol = _Symbol(match.group(1), "", rel)
                    index.setdefault(symbol.name, []).append(symbol)
        elif suffix == ".php":
            class_match = re.search(r"\bclass\s+([A-Za-z_]\w*)", text)
            scope = class_match.group(1) if class_match else ""
            for match in re.finditer(r"\bfunction\s+([A-Za-z_]\w*)\s*\(", text):
                symbol = _Symbol(match.group(1), scope, rel)
                index.setdefault(symbol.name, []).append(symbol)
    return index


def _resolve_handler(
    symbols: Dict[str, List[_Symbol]], name: str, scope: str = "",
) -> Tuple[str, str, str, str]:
    candidates = list(symbols.get(name, ()))
    if scope:
        scoped = [item for item in candidates if item.scope == scope.split("\\")[-1]]
        if scoped:
            candidates = scoped
    if len(candidates) == 1:
        item = candidates[0]
        return item.file_path, item.scope, item.label, "resolved"
    return "", scope.split("\\")[-1] if scope else "", "Function", (
        "ambiguous" if candidates else "unresolved"
    )


def _endpoint(
    *, project_id: str, framework: str, method: str, route: str, rel: str,
    line: int, handler: str, symbols: Dict[str, List[_Symbol]], scope: str = "",
) -> EndpointFact:
    handler_file, resolved_scope, label, status = _resolve_handler(symbols, handler, scope)
    normalized_route = _normalize_path(route)
    normalized_method = "ALL" if method.lower() in {"route", "use", "all", "any", "match"} else method.upper()
    return EndpointFact(
        endpoint_id=stable_id(project_id, framework, normalized_method, normalized_route, rel, line),
        project_id=project_id,
        framework=framework,
        http_method=normalized_method,
        path=normalized_route,
        file_path=rel,
        start_line=line,
        handler_name=handler,
        handler_scope=resolved_scope,
        handler_file=handler_file,
        handler_label=label,
        resolution_status=status,
        confidence=1.0 if status == "resolved" else 0.6,
    )


def analyze_project(
    root: Path | str,
    project_id: str,
    frameworks: Sequence[str],
    selected_paths: Optional[Sequence[str]] = None,
) -> WebAnalysisResult:
    root_path = Path(root).resolve()
    files = tuple(_source_files(root_path))
    symbols = _symbol_index(root_path, files)
    selected = {str(path).replace("\\", "/") for path in selected_paths or ()}
    framework_set = {str(value).strip().lower() for value in frameworks}
    endpoints: List[EndpointFact] = []
    for path in files:
        rel = _relative(root_path, path)
        if selected and rel not in selected:
            continue
        text = path.read_text(encoding="utf-8", errors="ignore")
        suffix = path.suffix.lower()
        if suffix == ".py" and {"fastapi", "django"} & framework_set:
            if "fastapi" in framework_set:
                for match in _FASTAPI_RE.finditer(text):
                    endpoints.append(_endpoint(
                        project_id=project_id, framework="fastapi", method=match.group("method"),
                        route=match.group("path"), rel=rel, line=_line(text, match.start()),
                        handler=match.group("handler"), symbols=symbols,
                    ))
            if "django" in framework_set:
                for match in _DJANGO_RE.finditer(text):
                    raw_handler = match.group("handler")
                    endpoints.append(_endpoint(
                        project_id=project_id, framework="django", method="ALL",
                        route=match.group("path"), rel=rel, line=_line(text, match.start()),
                        handler=raw_handler.split(".")[-1], symbols=symbols,
                    ))
        elif suffix in {".js", ".jsx"} and "express_js" in framework_set:
            for match in _EXPRESS_RE.finditer(text):
                endpoints.append(_endpoint(
                    project_id=project_id, framework="express_js", method=match.group("method"),
                    route=match.group("path"), rel=rel, line=_line(text, match.start()),
                    handler=match.group("handler"), symbols=symbols,
                ))
        elif suffix == ".php" and "laravel" in framework_set:
            for match in _LARAVEL_RE.finditer(text):
                endpoints.append(_endpoint(
                    project_id=project_id, framework="laravel", method=match.group("method"),
                    route=match.group("path"), rel=rel, line=_line(text, match.start()),
                    handler=match.group("handler"), scope=match.group("scope"), symbols=symbols,
                ))
    unique = {item.endpoint_id: item for item in endpoints}
    return WebAnalysisResult(
        project_id=project_id,
        endpoints=tuple(sorted(unique.values(), key=lambda item: (item.framework, item.file_path, item.start_line))),
    )


__all__ = ["WebAnalysisResult", "analyze_project"]

File: rust/crates/analyzer-web-overlays/src/web/pipeline.rs
//! Port `tools/web_framework/pipeline.py` — regex scan endpoints fastapi /
//! django / express / laravel + symbol index + resolution.

use std::collections::BTreeMap;
use std::path::Path;

use fancy_regex::Regex;

use crate::pyutil::{decode_utf8_replace, walk_sorted};
use crate::web::models::{dedupe_and_sort, EndpointFact, WebAnalysisResult};

const SKIP_DIRS: [&str; 11] = [
    ".git", ".cache", ".venv", "venv", "node_modules", "vendor", "dist", "build",
    "__pycache__", "bin", "obj",
];

pub struct WebRegexes {
    fastapi: Regex,
    django: Regex,
    express: Regex,
    laravel: Regex,
    py_def: Regex,
    py_class: Regex,
    js_function: Regex,
    js_arrow: Regex,
    php_class: Regex,
    php_function: Regex,
}

fn compile(pattern: &str) -> Regex {
    Regex::new(pattern).expect("static web overlay regex phải compile được")
}

impl WebRegexes {
    pub fn new() -> Self {
        Self {
            // _FASTAPI_RE — re.IGNORECASE | re.DOTALL
            fastapi: compile(concat!(
                r"(?is)@(?:\w+\.)?(?:app|router|api_router|blueprint)\.(?P<method>get|post|put|delete|patch|head|options|route|websocket)",
                r"\s*\(\s*[rRuUfF]*['\x22](?P<path>[^'\x22]+)['\x22][^)]*\)\s*",
                r"(?:async\s+)?def\s+(?P<handler>[A-Za-z_]\w*)\s*\(",
            )),
            // _DJANGO_RE — re.MULTILINE
            django: compile(concat!(
                r"(?m)\b(?:path|re_path)\s*\(\s*[rRuU]*['\x22](?P<path>[^'\x22]+)['\x22]\s*,\s*",
                r"(?P<handler>[A-Za-z_]\w*(?:\.[A-Za-z_]\w*)*)(?:\.as_view\s*\(\s*\))?",
            )),
            // _EXPRESS_RE — re.IGNORECASE
            express: compile(concat!(
                r"(?i)\b(?:app|router|server|api)\.(?P<method>get|post|put|delete|patch|head|options|all|use)",
                r"\s*\(\s*['\x22](?P<path>[^'\x22]+)['\x22]\s*,\s*(?P<handler>[A-Za-z_$][\w$]*)",
            )),
            // _LARAVEL_RE — re.IGNORECASE
            laravel: compile(concat!(
                r"(?i)\bRoute::(?P<method>get|post|put|delete|patch|options|any|match)\s*\(\s*",
                r"['\x22](?P<path>[^'\x22]+)['\x22]\s*,\s*\[\s*(?P<scope>[A-Za-z_\\][\w\\]*)::class",
                r"\s*,\s*['\x22](?P<handler>[A-Za-z_]\w*)['\x22]\s*\]",
            )),
            // symbol index — python
            py_def: compile(r"(?m)^\s*(?:async\s+)?def\s+([A-Za-z_]\w*)\s*\("),
            py_class: compile(r"(?m)^\s*class\s+([A-Za-z_]\w*)\b"),
            // symbol index — js
            js_function: compile(
                r"(?m)^\s*(?:export\s+)?(?:async\s+)?function\s+([A-Za-z_$][\w$]*)\s*\(",
            ),
            js_arrow: compile(
                r"(?m)^\s*(?:export\s+)?(?:const|let|var)\s+([A-Za-z_$][\w$]*)\s*=\s*(?:async\s*)?\([^)]*\)\s*=>",
            ),
            // symbol index — php
            php_class: compile(r"\bclass\s+([A-Za-z_]\w*)"),
            php_function: compile(r"\bfunction\s+([A-Za-z_]\w*)\s*\("),
        }
    }
}

impl Default for WebRegexes {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone)]
struct Symbol {
    scope: String,
    file_path: String,
    label: &'static str,
}

/// `pathlib.Path.suffix.lower()`.
fn suffix_lower(path: &Path) -> String {
    match path.extension() {
        Some(extension) => extension.to_string_lossy().to_lowercase(),
        None => String::new(),
    }
}

/// `_source_files` — rglob, bỏ phần tử skip trong relative parts, suffix
/// thuộc {.py, .js, .jsx, .php}. Walk sorted để deterministic (Python rglob
/// theo OS order nhưng mọi consumer cuối đều sort/stable).
fn source_files(root: &Path) -> Vec<(std::path::PathBuf, String)> {
    walk_sorted(root)
        .into_iter()
        .filter(|(path, rel)| {
            if rel.split('/').any(|part| SKIP_DIRS.contains(&part)) {
                return false;
            }
            matches!(suffix_lower(path).as_str(), "py" | "js" | "jsx" | "php")
        })
        .collect()
}

/// `_line` — `text.count("\n", 0, offset) + 1`.
fn line_at(text: &str, offset: usize) -> i64 {
    text.as_bytes()[..offset]
        .iter()
        .filter(|byte| **byte == b'\n')
        .count() as i64
        + 1
}

/// `_normalize_path`.
fn normalize_path(value: &str) -> String {
    let mut path = value.trim().to_string();
    if !path.starts_with('/') {
        path = format!("/{path}");
    }
    let mut normalized = String::with_capacity(path.len());
    let mut previous_slash = false;
    for character in path.chars() {
        if character == '/' {
            if previous_slash {
                continue;
            }
            previous_slash = true;
        } else {
            previous_slash = false;
        }
        normalized.push(character);
    }
    normalized
}

/// `_symbol_index`.
fn symbol_index(
    files: &[(std::path::PathBuf, String)],
    regexes: &WebRegexes,
) -> BTreeMap<String, Vec<Symbol>> {
    let mut index: BTreeMap<String, Vec<Symbol>> = BTreeMap::new();
    for (path, rel) in files {
        let bytes = match std::fs::read(path) {
            Ok(bytes) => bytes,
            Err(_) => continue, // read_text failed — Python ném lỗi, caller crash; không xảy ra trên corpus parity
        };
        let text = decode_utf8_replace(&bytes);
        let suffix = suffix_lower(path);
        let mut push = |name: &str, scope: &str, label: &'static str| {
            index.entry(name.to_string()).or_default().push(Symbol {
                scope: scope.to_string(),
                file_path: rel.clone(),
                label,
            });
        };
        if suffix == "py" {
            for capture in regexes.py_def.captures_iter(&text).flatten() {
                push(&capture[1], "", "Function");
            }
            for capture in regexes.py_class.captures_iter(&text).flatten() {
                push(&capture[1], "", "Class");
            }
        } else if suffix == "js" || suffix == "jsx" {
            for capture in regexes.js_function.captures_iter(&text).flatten() {
                push(&capture[1], "", "Function");
            }
            for capture in regexes.js_arrow.captures_iter(&text).flatten() {
                push(&capture[1], "", "Function");
            }
        } else if suffix == "php" {
            let scope = regexes
                .php_class
                .captures(&text)
                .ok()
                .flatten()
                .map(|capture| capture[1].to_string())
                .unwrap_or_default();
            for capture in regexes.php_function.captures_iter(&text).flatten() {
                push(&capture[1], &scope, "Function");
            }
        }
    }
    index
}

/// `_resolve_handler` — trả (handler_file, resolved_scope, label, status).
fn resolve_handler(
    symbols: &BTreeMap<String, Vec<Symbol>>,
    name: &str,
    scope: &str,
) -> (String, String, String, String) {
    let mut candidates: Vec<Symbol> = symbols.get(name).cloned().unwrap_or_default();
    if !scope.is_empty() {
        let tail = scope.rsplit('\\').next().unwrap_or("").to_string();
        let scoped: Vec<Symbol> = candidates
            .iter()
            .filter(|item| item.scope == tail)
            .cloned()
            .collect();
        if !scoped.is_empty() {
            candidates = scoped;
        }
    }
    if candidates.len() == 1 {
        let item = &candidates[0];
        return (
            item.file_path.clone(),
            item.scope.clone(),
            item.label.to_string(),
            "resolved".into(),
        );
    }
    let scope_tail = if scope.is_empty() {
        String::new()
    } else {
        scope.rsplit('\\').next().unwrap_or("").to_string()
    };
    let status = if candidates.is_empty() { "unresolved" } else { "ambiguous" };
    (String::new(), scope_tail, "Function".into(), status.into())
}

/// `_endpoint`.
#[allow(clippy::too_many_arguments)]
fn endpoint(
    project_id: &str,
    framework: &str,
    method: &str,
    route: &str,
    rel: &str,
    line: i64,
    handler: &str,
    symbols: &BTreeMap<String, Vec<Symbol>>,
    scope: &str,
) -> EndpointFact {
    let (handler_file, resolved_scope, label, status) =
        resolve_handler(symbols, handler, scope);
    let normalized_route = normalize_path(route);
    let normalized_method = if matches!(
        method.to_lowercase().as_str(),
        "route" | "use" | "all" | "any" | "match"
    ) {
        "ALL".to_string()
    } else {
        method.to_uppercase()
    };
    EndpointFact {
        endpoint_id: crate::web::models::stable_id(&[
            project_id.to_string(),
            framework.to_string(),
            normalized_method.clone(),
            normalized_route.clone(),
            rel.to_string(),
            line.to_string(),
        ]),
        project_id: project_id.to_string(),
        framework: framework.to_string(),
        http_method: normalized_method,
        path: normalized_route,
        file_path: rel.to_string(),
        start_line: line,
        handler_name: handler.to_string(),
        handler_scope: resolved_scope,
        handler_file,
        handler_label: label,
        resolution_status: status.clone(),
        confidence: if status == "resolved" { 1.0 } else { 0.6 },
    }
}

/// `analyze_project` — frameworks là list đã lowercase ("fastapi", "django",
/// "express_js", "laravel").
pub fn analyze_project(
    root: &Path,
    project_id: &str,
    frameworks: &[&str],
    selected_paths: &[String],
) -> WebAnalysisResult {
    let regexes = WebRegexes::new();
    let files = source_files(root);
    let symbols = symbol_index(&files, &regexes);
    let selected: std::collections::HashSet<&String> = selected_paths.iter().collect();
    let framework_set: std::collections::HashSet<&str> = frameworks.iter().copied().collect();
    let mut endpoints: Vec<EndpointFact> = Vec::new();
    for (path, rel) in &files {
        if !selected.is_empty() && !selected.contains(rel) {
            continue;
        }
        let bytes = match std::fs::read(path) {
            Ok(bytes) => bytes,
            Err(_) => continue,
        };
        let text = decode_utf8_replace(&bytes);
        let suffix = suffix_lower(path);
        if suffix == "py" && (framework_set.contains("fastapi") || framework_set.contains("django"))
        {
            if framework_set.contains("fastapi") {
                for capture in regexes.fastapi.captures_iter(&text).flatten() {
                    let whole = capture.get(0).map(|m| m.start()).unwrap_or(0);
                    endpoints.push(endpoint(
                        project_id,
                        "fastapi",
                        capture.name("method").map(|m| m.as_str()).unwrap_or(""),
                        capture.name("path").map(|m| m.as_str()).unwrap_or(""),
                        rel,
                        line_at(&text, whole),
                        capture.name("handler").map(|m| m.as_str()).unwrap_or(""),
                        &symbols,
                        "",
                    ));
                }
            }
            if framework_set.contains("django") {
                for capture in regexes.django.captures_iter(&text).flatten() {
                    let whole = capture.get(0).map(|m| m.start()).unwrap_or(0);
                    let mut raw_handler = capture.name("handler").map(|m| m.as_str()).unwrap_or("");
                    if let Some(stripped) = raw_handler.strip_suffix(".as_view") {
                        raw_handler = stripped;
                    }
                    let handler = raw_handler.rsplit('.').next().unwrap_or("");
                    endpoints.push(endpoint(
                        project_id,
                        "django",
                        "ALL",
                        capture.name("path").map(|m| m.as_str()).unwrap_or(""),
                        rel,
                        line_at(&text, whole),
                        handler,
                        &symbols,
                        "",
                    ));
                }
            }
        } else if (suffix == "js" || suffix == "jsx") && framework_set.contains("express_js") {
            for capture in regexes.express.captures_iter(&text).flatten() {
                let whole = capture.get(0).map(|m| m.start()).unwrap_or(0);
                endpoints.push(endpoint(
                    project_id,
                    "express_js",
                    capture.name("method").map(|m| m.as_str()).unwrap_or(""),
                    capture.name("path").map(|m| m.as_str()).unwrap_or(""),
                    rel,
                    line_at(&text, whole),
                    capture.name("handler").map(|m| m.as_str()).unwrap_or(""),
                    &symbols,
                    "",
                ));
            }
        } else if suffix == "php" && framework_set.contains("laravel") {
            for capture in regexes.laravel.captures_iter(&text).flatten() {
                let whole = capture.get(0).map(|m| m.start()).unwrap_or(0);
                endpoints.push(endpoint(
                    project_id,
                    "laravel",
                    capture.name("method").map(|m| m.as_str()).unwrap_or(""),
                    capture.name("path").map(|m| m.as_str()).unwrap_or(""),
                    rel,
                    line_at(&text, whole),
                    capture.name("handler").map(|m| m.as_str()).unwrap_or(""),
                    &symbols,
                    capture.name("scope").map(|m| m.as_str()).unwrap_or(""),
                ));
            }
        }
    }
    // sort + dedupe theo endpoint_id (last wins) rồi (framework, file, line).
    WebAnalysisResult {
        project_id: project_id.to_string(),
        endpoints: dedupe_and_sort(endpoints),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_path_collapses_slashes() {
        assert_eq!(normalize_path("items/{id}/"), "/items/{id}/");
        assert_eq!(normalize_path("//a///b"), "/a/b");
        assert_eq!(normalize_path("/x"), "/x");
    }

    #[test]
    fn fastapi_regex_matches_decorator_then_def() {
        let regexes = WebRegexes::new();
        let text = "@app.get(\"/items\")\nasync def read_items():\n    pass\n";
        let capture = regexes.fastapi.captures(text).unwrap().unwrap();
        assert_eq!(capture.name("method").unwrap().as_str(), "get");
        assert_eq!(capture.name("path").unwrap().as_str(), "/items");
        assert_eq!(capture.name("handler").unwrap().as_str(), "read_items");
    }

    #[test]
    fn laravel_regex_matches_class_handler() {
        let regexes = WebRegexes::new();
        let text = "Route::get('/users', [UserController::class, 'index']);\n";
        let capture = regexes.laravel.captures(text).unwrap().unwrap();
        assert_eq!(capture.name("method").unwrap().as_str(), "get");
        assert_eq!(capture.name("scope").unwrap().as_str(), "UserController");
        assert_eq!(capture.name("handler").unwrap().as_str(), "index");
    }
}
